Expand SSRF support in IAST to apache-httpclient, commons-httpclient and okhttp

Mariovido opened the pull request:
# What Does This Do
This changes the way of detecting an **SSRF** in the http client. In this PR we centralize inside the `HttpClientDecorator` the detection of the **SSRF** vulnerability. For now, we only have swapped the libraries that we were supporting with the previous approach (`commons-httpclient`, `apache-httpclient` and `okHttp`). The objective is to implement with this approach the rest of the clients supported by the `HttpClientDecorator`.

Even after centralizing the detection inside the `HttpClientDecorator` we needed to make some instrumentation to ensure the propagation.

# Motivation
With this change we want to expand the support for **SSRF** in the different clients supported by the `HttpClientDecorator`.

# Additional Notes
There are some cases where we cannot use this approach, so we need to maintain the previous approach and instrument the required methods to cover those cases.

# Contributor Checklist

- [x] Format the title [according the contribution guidelines](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#title-format)
- [x] Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any usefull labels](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#labels)
- [ ] Squash your commits prior merging or merge using GitHub's [Squash and merge](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits)
- [x] Don't use `close`, `fix` or any [linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue.  
  Use `solves` instead, and assign the PR [milestone](https://github.com/DataDog/dd-trace-java/milestones) to the issue
- [x] Update the [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) in case of new configuration flag or behavior

Jira ticket: [APPSEC-55237]

<!--
# Opening vs Drafting a PR:
When opening a pull request, please open it as a draft to not auto assign reviewers before you feel the pull request is in a reviewable state.

# Linking a JIRA ticket:
Please link your JIRA ticket by adding its identifier between brackets (ex [PROJ-IDENT]) in the PR description, not the title.
This requirement only applies to Datadog employees.
-->


[APPSEC-55237]: https://datadoghq.atlassian.net/browse/APPSEC-55237?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

manuel-alvarez-alvarez commented on dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpClientDecorator.java:
```diff
@@ -168,4 +173,23 @@ public long getResponseContentLength(final RESPONSE response) {
+
+  /* This method must be override after making the proper propagations to the client before **/
+  protected String sourceUrl(REQUEST request) throws URISyntaxException {
+    return "";
```
I think it's better to return a `null` here and short-circuit in `ssrfIastCheck`

Mariovido replied:
Changed the method to return an object

Mariovido commented on dd-java-agent/instrumentation/apache-httpclient-4/src/main/java/datadog/trace/instrumentation/apachehttpclient/ApacheHttpClientDecorator.java:
```diff
@@ -34,6 +36,16 @@ protected URI url(final HttpUriRequest request) {
   }
 
+  @Override
+  protected String sourceUrl(final HttpUriRequest request) {
```
Changed the override methods to return the tainted objects (URL, Strings, URI or Taintable Objects)

jandro996 commented on dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpClientDecorator.java:
```diff
@@ -168,4 +174,19 @@ public long getResponseContentLength(final RESPONSE response) {
     return 0;
   }
+
+  /* This method must be override after making the proper propagations to the client before **/
```
Typo?
```suggestion
  /* This method must be overridden after making the proper propagations to the client before **/
```

Mariovido replied:
Fixed

jandro996 commented on dd-java-agent/agent-bootstrap/src/test/groovy/datadog/trace/bootstrap/instrumentation/decorator/HttpClientDecoratorTest.groovy:
It would be nice to have a unit test case for the client decorator that matches call to the SSRFModule

Mariovido replied:
I've added a unit test where we test if we are calling the `ssrfModule` when iast is fully activated

jandro996 commented:
Correct me if I'm wrong but I think that SsrfModule#onURLConnection(@Nullable String url, @Nullable Object host, @Nullable Object uri) is not used anymore with these changes, so it's better to remove it from the interface and the implementation to avoid dead code

Mariovido commented:
> Correct me if I'm wrong but I think that SsrfModule#onURLConnection(@nullable String url, @nullable Object host, @nullable Object uri) is not used anymore with these changes, so it's better to remove it from the interface and the implementation to avoid dead code

I have removed the method and added an exclusion for the coverage. It seems that method was the only one that tested the `EvidenceBuilder`. I have created a JIRA ticket to solve this problem later. For now, the method will be removed and we exclude `EvidenceBuilder` from the coverage report

manuel-alvarez-alvarez commented on dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/decorator/HttpClientDecorator.java:
```diff
@@ -168,4 +174,19 @@ public long getResponseContentLength(final RESPONSE response) {
+    if (InstrumenterConfig.get().getIastActivation() != ProductActivation.FULLY_ENABLED) {
+      return;
+    }
+    final SsrfModule ssrfModule = InstrumentationBridge.SSRF;
```
I'll rather short circuit first if the actual URL is null (`sourceUrl(request)`)

Mariovido replied:
Done

manuel-alvarez-alvarez commented on dd-java-agent/agent-iast/src/main/java/com/datadog/iast/sink/SsrfModuleImpl.java:
```diff
@@ -22,64 +18,4 @@ public void onURLConnection(@Nullable final Object url) {
     }
     checkInjection(VulnerabilityType.SSRF, url);
   }
-
```
Nice, this was a bit ugly in the first place :smile:

manuel-alvarez-alvarez commented on dd-smoke-tests/iast-util/src/testFixtures/groovy/datadog/smoketest/AbstractIastSpringBootTest.groovy:
```diff
@@ -712,6 +712,43 @@ abstract class AbstractIastSpringBootTest extends AbstractIastServerSmokeTest {
     'host'    | 'dd.datad0g.com'
   }
 
+  void 'ssrf is present (different clients)'() {
```
Nice!

Mariovido replied:
Changed to include the path

manuel-alvarez-alvarez commented on internal-api/src/main/java/datadog/trace/bootstrap/instrumentation/api/URIUtils.java:
Will it make sense to apply call sites to this class (you will get all the required propagation)?

jandro996 commented:
LGTM! but please check the build, it seems that datadog.trace.api.iast.util.PropagationUtils is not passing the test coverage job
